3단계 - 구간 추가 기능

mukeunzi opened the pull request:
준홍님 안녕하세요! 
3단계 PR요청 드립니다 😃

[2단계 미션에서 리뷰남겨주신 것](https://github.com/next-step/atdd-subway-admin/pull/817)도 함께 반영했습니다. 확인 부탁드립니다!
항상 감사합니다 :)

mukeunzi committed:
docs(docs): Step3 기능 요구사항 작성

mukeunzi committed:
refactor(domain): 접근제어자 추가/수정

- 불필요한 인스턴스 생성 방지를 위해 생성자함수 접근제어자 protected로 변경
- 컨트롤러에 선언된 서비스 멤버변수 불변성을 위해 final 접근제어자 추가

mukeunzi committed:
refactor(domain): Line 생성요청 시, distance도 입력받도록 수정

mukeunzi committed:
feat(domain): 지하철역 구간추가 기능 구현

- 성공 케이스에 대한 인수테스트 작성
- Section 엔티티 생성

mukeunzi committed:
feat(domain): Sections 일급컬렉션 생성

- Line, Section 생성 시 Sections에 항목 추가

mukeunzi committed:
feat(domain): 상행역, 하행역이 이미 노선에 등록되어있을 경우 예외처리

- 위 조건일 경우 400 Error 발생
- 공통 예외처리를 위한 CommonExceptionHandler 작성
- 상행역, 하행역 등록여부 확인을 위한 검증메소드 추가

mukeunzi committed:
refactor(domain): Section 검증 메소드 메소드 구체화

mukeunzi committed:
feat(domain): 상행역 하행역 둘 중 하나에도 포함되어있지 않으면 구간추가 불가 예외처리

- 구간추가 역이 기존 노선의 역에 포함되어있는지 검증하는 메서드 구현
- 예외처리를 위한 InvalidSectionException 클래스 추가
- 인수테스트 작성

mukeunzi committed:
fix(domain): 상행역, 하행역에 구간 추가 시 검증하는 조건 변경

- 구간추가 역이 기존 노선 상행역 하행역 어느 곳에 포함되어 있어도 유효성 검증 통과하도록 수정

mukeunzi committed:
refactor(domain): 람다식에서 메소드 호출 시 method reference 방식(::)으로 변경

mukeunzi committed:
refactor(domain): Line생성 시, sections에 구간을 추가하는 시점 변경

mukeunzi committed:
refactor(domain): Sections에 CASCADE 설정 추가

- Station과 Line의 중간 테이블 역할인 Section의 List를 담는 Sections에 CASCADE설정
- CASCADE를 설정함으로써 SectionRepository 삭제

mukeunzi committed:
feat(domain): 사이에 구간을 추가할 경우 거리에 대한 예외처리 추가

- 예외처리를 위한 InvalidDistanceException 클래스 생성
- 사이에 구간추가 시, 기존 거리보다 거리가 짧은지 검증하는 메소드 추가
- 완료항목 업데이트

mukeunzi committed:
refactor(domain): Station equals 메서드 구현

mukeunzi committed:
fix(domain): 노선 생성 시에만 Section을 추가하기 위해 구간추가 시점 변경

mukeunzi committed:
docs(docs): 완료항목 업데이트

Junhong-Kim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -31,6 +31,7 @@ public LineResponse saveLine(LineRequest lineRequest) throws NotFoundStationExce
         Station downStation = stationService.findStationById(lineRequest.getDownStationId());
 
         Line persistLine = lineRepository.save(lineRequest.toLine(upStation, downStation));
+        persistLine.addSection();
```
메서드 내부에서 Section을 생성하고 있는데요.
`addSection()` 인자로 생성될 Section을 넘겨주는 방식으로 바꾸어 어떤 Section이 생성되는지 알 수 있으면 좋을 것 같아요.

Junhong-Kim commented on src/main/java/nextstep/subway/application/SectionService.java:
```diff
@@ -0,0 +1,32 @@
+        Station upStation = stationService.findStationById(sectionRequest.getUpStationId());
+        Station downStation = stationService.findStationById(sectionRequest.getDownStationId());
+
+        Section newSection = sectionRequest.toSection(line, upStation, downStation);
```
모든 인자를 새로받아서 `SectionRequest`가 Section을 만들고 있네요.
이는 `Section` 클래스가 인스턴스를 생성해서 반환하는 역할인 것 같습니다.

mukeunzi replied:
`toSection()` 가 Section 클래스에 있어야 할 것 같다는 의견으로 이해했는데 맞을까요? 
저는 sectionRequest를 통해 Section 생성을 위한 값을 입력받고, 이를 DB에 insert하기 전에 이 서비스 로직에서 *Id 로 입력받은 값을 각각의 인스턴스로 변환하여, Section을 생성하기 때문에 이는 SectionRequest의 역할이라고 생각했었습니다! 

만약 toSection을 Section으로 옮기게 되면, `Section.toSection(line, upStation, downStation, sectionRequest.getDistance());` 와 같이 distance를 가져오기 위해 getter를 써야하는데, 그렇게 하는 것이 맞을까요?

Junhong-Kim replied:
현재는 Reqeust 객체가 Service에서 새로운 값을 받아 Section 인스턴스를 반환하고 있어요.
Request 객체는 Controller에서 요청 값을 받아 요청 값에 대한 유효성 검사하는 책임만 갖고,

`Section` 클래스에서 인스턴스 생성에 대한 역할을 하면 좋을 것 같아요.
왜냐하면 생성자에서 인스턴스 생성에 대한 유효성 검사를 포함할 수 있기 때문입니다.
정적 패토리 메서드와 관련해서 참고할만한 [링크](https://tecoble.techcourse.co.kr/post/2020-05-26-static-factory-method/)도 공유드립니다.

> 만약 toSection을 Section으로 옮기게 되면, Section.toSection(line, upStation, downStation, sectionRequest.getDistance()); 와 같이 distance를 가져오기 위해 getter를 써야하는데, 그렇게 하는 것이 맞을까요?

Request 객체로 부터 값을 얻어야하는 부분이기 때문에 말씀해주신 것처럼 수정하셔도 괜찮습니다.

mukeunzi replied:
넵 준홍님! 상세한 설명 감사합니다 🙇 

객체의 책임과 연관있던 것이었군요. 
미션을 진행하면서 getter를 사용하지 않고 객체에게 메시지를 보내는 것에 대해서는 훈련이 되었는데, 책임을 구분하는 것에 대해서는 부족했던 것 같습니다.ㅠㅠ
앞으로 미션 진행 시 책임에 대해서도 함께 고민할 수 있도록 하겠습니다 !

mukeunzi replied:
코드 상에 Line에도 정적팩토리 메서드 수정이 필요하여 함께 적용하였습니다~!

Junhong-Kim commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,102 @@
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "down_station_id")
+    private Station downStation;
+    private Long distance;
```
원시 값을 포장해보면 좋을 것 같아요.

Junhong-Kim replied:
[코멘트](https://github.com/next-step/atdd-subway-admin/pull/824/files#r903800062) 확인 부탁드립니다 😊

mukeunzi replied:
넵 반영했습니다!

Junhong-Kim commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    private boolean isBetweenSection(Line line, Station newUpStation, Station newDownStation) {
+        if (line.getUpStation().equals(newUpStation)) {
```
객체에 메시지를 보내는 방식으로 수정해보면 좋을 것 같아요.

mukeunzi replied:
앗 이부분을 놓쳤네요. 반영하겠습니다!

Junhong-Kim commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    public Section(Line line, Station upStation, Station downStation, Long distance) {
+        validateIncludeAnyStation(line, upStation, downStation);
```
Section 전체에서 Station이 존재하는지 검증하도록 수정하면 좋을 것 같아요.

mukeunzi replied:
`Sections.java`에 `validateUnique()`에서 해당 로직 실행되고 있습니다!

Junhong-Kim replied:
`Line`에 `Section`을 추가할 때 유효성 검사를 하고 있는데 어떤 의도의 validate일까요?

mukeunzi replied:
`validateUnique()` 를 통해 이미 노선에 등록된 sections 들과 중복되는 것이 없는지 검증하고,
`validateDistance()` 를 통해 구간 사이에 추가되는 노선일 경우 distance가 유효한지를 검증하고 있습니다!

Junhong-Kim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,37 @@
+
+    private void validateUnique(Section newSection) {
+        sections.forEach(section -> {
+            boolean isExists = section.isExistsSection(newSection);
+            if (isExists) {
+                throw new DuplicatedSectionException(newSection.getUpStation().getId(), newSection.getDownStation().getId());
+            }
```
새롭게 추가되는 구간에 대한 검증을 `for`문을 돌면서 확인하고 있네요.
Sections에 메세지를 보내어 확인하면 될 것 같습니다.

mukeunzi replied:
해당 방법에 대해 추가적으로 설명해주실 수 있을까요? 
이 클래스가 Sections인데, 어떻게 메시지를 보내야 하는지 잘 모르겠습니다..ㅜ

Junhong-Kim replied:
다음과 같은 방법을 활용해보면 좋을 것 같아요 😊
```java
if (sections.contains(newSection)) {
    // do something
}
```

mukeunzi replied:
앗 넵! 반영하겠습니다!

Junhong-Kim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.List;
+
+@Embeddable
+public class Sections {
```
일급 컬렉션 사용 👍

Junhong-Kim commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,102 @@
+        return distance;
+    }
+
+    public boolean isExistsSection(Section newSection) {
```
이 부분도 Sections의 역할로 옮기는게 좋을 것 같아요.

mukeunzi replied:
Sections.java에서 sections를 iteration을 돌면서 실행되기 때문에 Section에 둔 것인데, 어떤 이유 때문에 Sections가 더 적합하다고 생각하시는지 궁금합니다!

Junhong-Kim replied:
다음 코멘트를 확인해주세요!
https://github.com/next-step/atdd-subway-admin/pull/824#discussion_r905136292

mukeunzi replied:
링크 남겨주신 코드와 함께 반영했습니다! 🙇

Junhong-Kim commented on src/main/java/nextstep/subway/exception/CommonExceptionHandler.java:
```diff
@@ -0,0 +1,27 @@
+
+@RestControllerAdvice
+public class CommonExceptionHandler {
+    @ExceptionHandler(DuplicatedSectionException.class)
```
`BAD_REQUEST`를 응답하는 핸들러들을 하나의 핸들러에서 처리하도록 수정해보면 좋을 것 같아요.
```suggestion
    @ExceptionHandler({FileSystemException.class, InvalidSectionException.class, InvalidDistanceException.class})
```

mukeunzi replied:
앗 넵 확인 후 반영하겠습니다!

Junhong-Kim commented on src/main/java/nextstep/subway/exception/CommonExceptionHandler.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class CommonExceptionHandler {
```
피드백 반영 👍

Junhong-Kim commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -115,6 +117,110 @@ void deleteLine() {
+     * When 지하철노선에 구간을 추가하면
+     * Then 노선에 구간이 등록된다.
+     */
+    @DisplayName("지하철노선에 구간을 추가한다.")
```
새로운 구간을 `상행 종점/하행 종점/역 사이`에 추가할 때를 각각 테스트 코드로 작성하면 좋을 것 같아요.

mukeunzi replied:
넵! 반영하겠습니다 :)

Junhong-Kim commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    private void validateIncludeAnyStation(Line line, Station newUpStation, Station newDownStation) {
+        if (line.getUpStation().equals(newUpStation) || line.getUpStation().equals(newDownStation)) {
+            return;
+        }
+
+        if (line.getDownStation().equals(newUpStation) || line.getDownStation().equals(newDownStation)) {
+            return;
+        }
```
`getter` 사용을 지양하는 코드로 수정해보면 좋을 것 같아요.
그리고 `if`문 안의 로직을 의미있는 메서드로 추출하면 가독성을 높일 수 있습니다.

mukeunzi replied:
넵! 이부분도 같이 수정하겠습니다 :)

Junhong-Kim reviewed (changes requested):
3단계 미션 잘 구현해주셨습니다 👍
리뷰 내용 확인하시고 다시 요청해주시면 감사드립니다!

mukeunzi committed:
test(domain): 노선 추가 성공케이스 구분하여 테스트케이스 작성

mukeunzi committed:
refactor: 같은 에러 상태코드를 반환하는 클래스를 하나의 핸들러에서 공통으로 처리하도록 리팩토링

mukeunzi committed:
refactor(domain): 조건문에서 getter 메소드 사용하는 로직 제거

- 객체한테 메시지를 보내도록 수정

mukeunzi committed:
refactor(domain): 노선생성 과정에서 구간 추가 시, 구간 추가되는 Section을 파라미터로 넘기도록 수정

Junhong-Kim commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -32,8 +32,8 @@ public Line(String name, String color, Long distance, Station upStation, Station
+
+    public void addSection(Station upStation, Station downStation, Long distance) {
         this.sections.add(new Section(this, upStation, downStation, distance));
     }
```
다음과 같은 메서드를 사용해서 메서드명 의미를 좀 더 명확하게 하면 좋을 것 같아요.
```java
public void addSection(Section section) {
    // ...
}
```

mukeunzi replied:
반영했습니다!

Junhong-Kim commented on src/main/java/nextstep/subway/ui/LineController.java:
```diff
@@ -49,4 +50,10 @@ public ResponseEntity<LineResponse> deleteLine(@PathVariable Long id) {
+
+    @PostMapping("/{lineId}/sections")
+    public ResponseEntity<SectionResponse> createSection(@PathVariable Long lineId, @RequestBody SectionRequest sectionRequest) throws NotFoundStationException {
+        SectionResponse section = sectionService.addSection(lineId, sectionRequest);
```
`Line`에 `Section`을 추가하는 것이므로 `LineService`에 `addSection`을 구현하면 좋을 것 같아요.

mukeunzi replied:
다른 코멘트와 함께 반영했습니다!

Junhong-Kim commented on src/main/java/nextstep/subway/ui/LineController.java:
```diff
@@ -49,4 +50,10 @@ public ResponseEntity<LineResponse> deleteLine(@PathVariable Long id) {
     }
+
+    @PostMapping("/{lineId}/sections")
+    public ResponseEntity<SectionResponse> createSection(@PathVariable Long lineId, @RequestBody SectionRequest sectionRequest) throws NotFoundStationException {
```
`LineController`에 다른 API를 보시면 아시겠지만 `SectionResponse`가 아닌 `LineResponse`를 사용하면 될 것 같아요.

mukeunzi replied:
넵 요구사항에도 LineResponse로 응답하도록 되어있는 것 확인했습니다. 👍 

적용하다 보니 SectionResponse, SectionService가 불필요하다고 판단해서 삭제했고, 이에 따라 서비스 단의 addSection 메서드를 LineService로 옮겼습니다. :)

Junhong-Kim reviewed (changes requested):
피드백 잘 반영해주셨는데 좀 더 수정해볼 만한 내용 코멘트 남겼습니다!
그리고 PR로 질문 남겨주신 사항들은 코멘트로 답변드렸습니다.
리뷰 내용 확인하시고 다시 요청해주시면 감사드립니다. ☺️

mukeunzi committed:
refactor(domain): 메서드 선언 위치 변경

- getter가 가장 마지막에 선언대도록 순서 변경

mukeunzi committed:
refactor(domain): 불필요하게 구현된 메서드 제거(List 에서 지원하는 API로 대체)

mukeunzi committed:
fix: 구간추가 요청 시, 응답값 구조 변경

mukeunzi committed:
refactor(domain): 정적팩토리 메소드 도메인 클래스로 이동

mukeunzi committed:
fix(domain): Line entity 구조 변경 및 Line 생성 시 리턴되는 응답값 구조 변경

- Line에는 name, color만 컬럼으로 갖고, Sections는 별도의 테이블 값을 참조하는 구조로 변경
- Line 생성 시, sections 값을 List형식으로 응답하기 위한 StationSimpleResponse(dto) 생성

mukeunzi committed:
fix(domain): 지하철 노선 수정기능에 임시 주석 제거

mukeunzi committed:
fix(domain): 구간추가 시 상행역/하행역 하나라도 포함되어야 하는 예외처리 수정

- 구간추가 시에만 포함여부를 체크하도록 validation 실행 위치 수정

mukeunzi committed:
fix(domain): 사이에 구간추가 시 거리에 대한 예외처리 추가

mukeunzi committed:
fix(domain): 테스트코드 수정 및 사이에 구간이 추가될 경우 기존 section 업데이트하는 로직 추가

mukeunzi committed:
feat(domain): Distance 일급컬렉션 생성

mukeunzi committed:
refactor(domain): Line entity에서 사용하지 않는 컬럼 제거

mukeunzi commented:
준홍님 안녕하세요! 
리뷰내용을 반영하는 과정에서 고민이 길어지고, 요구사항을 잘못 구현했던 부분들이 있어 수정하느라 시간이 조금 오래 걸렸습니다 🥲

리뷰 남겨주신 내용은 모두 반영되었고, 아래 내용이 추가로 수정되었습니다.

- Line entity에 `upStationId, downStationId, distance` 필드 제거 
- 기존 노선 사이에 구간이 추가될 경우 기존 노선의 상행/하행역 및 distance 변경
  - ex) `[강남역, 잠실역, 10]` 노선/구간 추가 -> `[강남역, 역삼역, 2]` 구간 추가 -> 기존 노선 `[역삼역, 잠실역, 8]` 로 구간 변경

이미 한번 리뷰해주셨는데 큰 수정을 하게되어 죄송한 마음입니다 ㅠㅠ
커밋 별로 확인하는 것이 더 보기 편할 수 있을 것 같습니다..! 😢 

이번 리뷰 반영 과정에서는 인수테스트 보다 기능구현에 더 초점이 맞춰진 것 같기도 하지만, 큰 수정을 하는데 있어 인수테스트가 작성되어 있어 검증하기 훨씬 수월했고, 안전장치의 역할이 될 수 있다는 느낄 수 있었던 것 같습니다. 

그럼 리뷰 부탁드리겠습니다. 감사합니다! 🙌🏻

Junhong-Kim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -27,11 +27,8 @@ public void add(Section newSection) {
-            }
-        });
+        if (sections.contains(newSection)) {
+            throw new DuplicatedSectionException(newSection.getUpStation().getId(), newSection.getDownStation().getId());
```
`Section` 객체만 넘겨도 괜찮을 것 같아요!

mukeunzi replied:
넵! 반영하겠습니다

Junhong-Kim commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -33,6 +36,18 @@ public Line(String name, String color, Long distance, Station upStation, Station
+
+    public static Line of(LineUpdateRequest lineUpdateRequest, Long distance, Station upStation, Station downStation) {
+        return new Line(lineUpdateRequest.getName(), lineUpdateRequest.getColor(), distance, upStation, downStation);
+    }
```
정적 팩터리 메서드에 `Request` 객체를 매개변수로 받으면 `of` 메서드가 `Request` 객체에 종속됩니다.
`Request` 객체에서 값만 받으면 좋을 것 같아요.

mukeunzi replied:
앗 종속관계 제거하도록 하겠습니다

mukeunzi committed:
refactor(domain): 멤버변수에 final 선언 및 조건문 단순화

Junhong-Kim commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    public static Section of(Line line, Station upStation, Station downStation, Long distance) {
+        validateIncludeAnyStation(line, upStation, downStation);
```
`Section` 인스턴스 생성시의 `Line`을 인자로 받아 `Line`에 이미 존재하는 `Section`인지 체크하는 것이 아닌,
`Line`에 `Section`을 추가할 때 확인하면 될 것 같아요.

mukeunzi replied:
넵! Sections validation을 한곳에서 수행하는 것이 좋겠네요. 반영하겠습니다.

Junhong-Kim reviewed (approved):
피드백 잘 반영해주셨네요 👍
소소한 코멘트 남겼는데 다음 단계 시작 전 반영해보면 좋을 것 같아요.

이번 단계는 이만 merge 하겠습니다.
다음 단계도 화이팅입니다! ☺️
